File: WEB/backend/python/flask/database/mongodb/restapi.py
from flask import Flask,request,jsonify
from pymongo import MongoClient
from bson import ObjectId
app=Flask(__name__)
Client=MongoClient("mongodb://localhost:27107")
db=Client["users_db"]
users=db["users"]

# POST
@app.route("/register",methods=["POST"])
def register():
    data=request.json
    name=data["name"]
    email=data["email"]
    users.insert_one({
        "name":name,
        "email":email
    })
    return jsonify({
        "message":"registered successfully"
    })

# GET
@app.route("/users",methods=["GET"])
def Users():
   result=[]
   for user in users.find():
       user["_id"]=str(user["_id"])
       result.append(user)
       
   return jsonify(result)
    
# GET one user 
@app.route("/user/<id>",methods=["GET"])
def GetUser(id):
    user=users.find_one({
        "_id":ObjectId(id)
    })
    if not user:
        return jsonify({
            "message":"user not found"
        }),404
    user["_id"]=str(user["_id"])
    return jsonify(user)

# update user
@app.route("/update/<id>",methods=["PUT"])
def update_user(id):
    data=request.json
    user=users.find_one(
        {"_id":ObjectId(id)},
        {
            "$set":
                {
                "name":data["name"],
                "email":data["email"]
                }
        }
    )
    return jsonify({
        "message":"user updated successfully"
    })

# delete user
@app.route("/delete/<id>",methods=["DELETE"])
def deleteuser(id):
    data=request.json
    user=user.delete_one(
        {"_id":ObjectId(id)}
    )
    return jsonify({
        "message":"deleted succesfully"
    })
if __name__=="_main_":
    app.run(debug=True,port=5001)

File: WEB/backend/javascript/express/7.databse integeration/mongoose/restapi.js
const express=require("express");
const app=express()
const mongoose=require("mongoose");
mongoose.connect("mongodb://localhost:27017/TestDb")
.then(()=>console.log("db is connected"))
.catch(err=>console.log(err))
const UserSchema=new mongoose.Schema({
    name:{type:String},
    email:{type:s=String},
    password:{type:String}
});
const User=mongoose.model("users",UserSchema);

// post
app.post("/register",async(req,res)=>{
    try{
        const {name,email,password}=req.body
        const user=await UserSchema.create({
            name,
            email,
            password
        })
        res.status(201).json({
            message:"registerd successfully",
            user
        })
    }catch(err){
        console.log(err);
        res.status(500).json({
            message:"server error"
        })
    }
})

//Get
app.get("/users",async(req,res)=>{
    try{
        const users=await User.find();
        res.status(200).json({
            message:"users found",
            user
        })
    }catch(err){
        console.log(err);
    }
})

app.get("/user/:id",async(req,res)=>{
    try{
        const {id}=req.params.id;
        const user=await user.findById(id);
        res.status(200).json({
            user
        })
    }catch(err){
        console.log(err);
    }
})
app.get("/user/:email",async(req,res)=>{
    try{
        const email=req.params.email
        const user=await User.find({email:email});
        res.status(200).json({
            user
        })
    }catch(err){
        console.log(err);
    }
})

//put 

// update all
app.put("/update",async(req,res)=>{
    try{
        const {name,email}=req.body
        const user=await user.updateMany(
            {},
            {$set:
                {
                    name:name,
                    email:email
                }
            }
        )
        res.status(201).json({
            user
        })
    }catch(err){
        console.log(err);
    }
})

// update one
app.put("/updateone/:email",async(req,res)=>{
    try{
        const email=req.params.email
        const {name,password}=req.body;
        const user=await User.updateOne(
            {email:email},
            {$set:{
                name:name,
                password:password
            }}
        )
        res.status(200).json({
            message:"updated successfully",
            user
        })
    }catch(err){
        console.log(err);
    }
})
// update by id
app.put("updateid/:id",async(req,res)=>{
    try{
        const {id}=req.params.id;
        const {email,password}=req.body;
        const user=await User.findByIdAndUpdate(
            id,
            {
                name:name,
                email:email
            },
            {new:true}
        )
    }catch(err){
        console.log(err)
    }
})
app.patch("/update/:id",async(req,res)=>{
    try{
        const id=req.params.id;
        const {email,password}=req.body;
        const user=await User.findByIdAndUpdate(id
            ,
            {
                email:email,
                passord:password
            }
            ,{
                new:true
            }
        )
    }catch(err){
        console.oog(err);
    }
})
//delete
app.delete("/delete/:email",async(req,res)=>{
    try{
        const email=req.params.email;
        const {name,password}=req.body;
        const user=await User.deleteOne({email:email});
        res.status(200).json({
            user
        })
    }catch(err){
        console.log(err)
    }
})

//delete by id
app.delete("deleteuser/:id",async(req,res)=>{
    try{
        const id=req.params.id;
        const user=await user.findByIdAndDelete(id);
        res.status(200).json({
            message:"deleted successfully"
        })
    }catch(err){
        console.log(err);
    }
})